Personal/mihansen/aad proxy

hansenms opened the pull request:
## Description
Introduces SMART on FHIR AAD Proxy

## Related issues
Addresses #222 

## Testing
E2E Tests using Selenium Added.

jmandel commented on src/Microsoft.Health.Fhir.Api/Controllers/AadProxyController.cs:
```diff
@@ -0,0 +1,363 @@
+            // TODO: Add handling of 'aud' -> 'resource'
+            if (grantType != "authorization_code")
+            {
+                List<KeyValuePair<string, string>> fields = new List<KeyValuePair<string, string>>();
```
Quick question: am I right in understanding this is a guaranteed failure condition?

If so, why not just return an "unsupported grant type" error? (And if not, which other grant types do we expect here?)

hansenms replied:
Actually, that is not a guaranteed failure. AAD supports client_credentials, etc. and we don't want to "disable" those features just because we enable SMART proxy. It would actually cause us a lot of problems if we did because many of the E2E tests obtain tokens that way. So this is sort of a bypass; if it is not authorization code, just pass on to AAD>

johnstairs replied:
I think we should change the tests to use an environment variable instead and make this case an error as Josh suggests. Not in this PR though...

hansenms replied:
@johnstairs, I agree, I will put a TODO note in the code to remove this. I think we have will some code changes related to multiple IDPs that will affect this too. So I will mark the code for now.

hansenms commented on samples/apps/SmartLauncher/appsettings.json:
```diff
@@ -0,0 +1,5 @@
+{
+    "FhirServerUrl": "https://localhost:44348",
+    "ClientId": "APP-ID",
+    "SmartAppUrl": "/sampleapp/launch.html"
```
Good point. I have changed it to `DefaultSmartAppUrl`, I was thinking of providing more of an array of possible apps and a dropdown box, but that would sort of limit it to what was defined in the app. Maybe room for some improvements down the line.

hansenms commented:
I think it is time to let the review process start on this. Some comments:

The AAD proxy works by:
1. Proxying the request to AADs `/authorize` endpoint. In that process, the `aud=` parameter is renamed to `resource=` and the `launch=` context parameter is appended to the `state=` parameter (which means it will be returned with authorization code. 
2. The Reply URL is changed from `https://myapp/reply` to `https://fhir-server/AadProxy/callback/<Base64(https://myapp/reply)>`
3. The authorization code is returned to the AAD proxy where the launch context is unpacked from the state parameter and a compound authorization code is created. This compound code contains both the authorization code and the launch context.
4. The AAD proxy sends the compund code to the SMART app reply URL, `https://myapp/reply` 
5. The SMART App sends the compound code to `https://fhir-server/AadProxy/token`.
6. The proxy unpacks the compound code and send sthe original authorization code to AAD, where it obtains a token.
7. Selected fields from the launch context are added to the token response and returned to the SMART App.

There will be a follow up PR to once all of this is merged. This PR, will: 
1. remove the default CORS settings (`https://localhost:6001`) from the template and add the CORS settings as template deployment parameter in the build system.
2. Disable the AAD proxy as a default setting and add it as a deployment parameter in the build systemt.

@johnstairs The launch context is now passed as part of the `state=` parameter in the authorization request. This removes the nasty wildcard reply URL which was previously required. I also experimented with URL encoding (instead of base64) the app reply URL in the proxy reply URL, but that doesn't work. However, current scheme *should* be B2C compatible, albeit with some readability and possible case sensitivity issues. Room for improvement.

johnstairs commented on src/Microsoft.Health.Fhir.Api/Controllers/AadProxyController.cs:
```diff
@@ -0,0 +1,355 @@
+        /// <param name="sessionState">session_state URL parameter.</param>
+        /// <param name="error">error URL parameter.</param>
+        /// <param name="errorDescription">error_description URL parameter.</param>
+        [HttpGet("callback/{encodedRedirect}")]
```
I'm not sure this will work properly if the base64 encoding contains a /.  It might be safer to adopt a url safe base64 encoding.

johnstairs replied:
Also, since they often end with =, this could be problematic when used in query strings.

hansenms replied:
I have switched to using `Base64UrlEncoder` throughout the code and the `New-FhirServerSmartClientReplyUrl` has been updated accordingly.

johnstairs commented on src/Microsoft.Health.Fhir.Api/Controllers/AadProxyController.cs:
```diff
@@ -0,0 +1,355 @@
+                {
+                    openIdConfigurationResponse = httpClient.GetAsync(new Uri(openIdConfigurationUrl)).GetAwaiter().GetResult();
+                }
+                catch (Exception ex)
```
We should avoid catching all exceptions types.

hansenms replied:
I adopted this pattern from `CapabilityStatementExtensions.cs`. There could be a few different things to catch. Do you propose we walk through all of them?

johnstairs replied:
HttpRequestException and OperationCanceledException should cover it. Also, I'd just do `openIdConfigurationResponse.EnsureSuccessStatusCode()` instead of the if block below.

johnstairs commented on src/Microsoft.Health.Fhir.Api/Controllers/AadProxyController.cs:
```diff
@@ -0,0 +1,355 @@
+                }
+            }
+
+            if (openIdConfigurationResponse.IsSuccessStatusCode)
```
And if we don't get success? Seems like we should throw as we do above.

johnstairs replied:
We could also wrap this in a Polly retry policy...

johnstairs commented on src/Microsoft.Health.Fhir.Api/Controllers/AadProxyController.cs:
```diff
@@ -0,0 +1,355 @@
+
+            newQueryString += $"&state={newState}";
+
+            return RedirectPermanent($"{_aadAuthorizeEndpoint}?{newQueryString}");
```
Is this supposed to be a permanent redirect?

hansenms replied:
Switched to `Redirect`

johnstairs commented on src/Microsoft.Health.Fhir.Api/Controllers/AadProxyController.cs:
```diff
@@ -0,0 +1,355 @@
+
+                EnsureArg.IsNotNull(scope, nameof(scope));
+                var scopes = scope.Split(' ');
+                string newScopes = string.Empty;
```
Should use a stringbuilder instead of repeated string concatenation

hansenms commented:
@johnstairs, I switched to `Base64UrlEncoder` and addressed most other issues. What are your thoughts on the `catch (Exception ec) {}` when trying to get the OpenIdConfiguration statement. What should we be catching?

hansenms commented:
@johnstairs, Also added a feature filter to deny access to the controller unless the feature is switched on. It could have been done simpler but it should do the trick. Let me know what you think.

johnstairs commented on samples/apps/SmartLauncher/.gitignore:
```diff
@@ -0,0 +1,2 @@
+bin/
+obj/
```
Is this file necessary? We already have this in the repo root.

johnstairs commented on samples/apps/SmartLauncher/Startup.cs:
```diff
@@ -0,0 +1,58 @@
+            }
+
+            app.UseDefaultFiles();
+            app.UseStaticFiles(new StaticFileOptions { FileProvider = new EmbeddedFileProvider(Assembly.GetExecutingAssembly(), "SmartLauncher.wwwroot") });
```
You don't really need the embedded file provider here. It could work just as well with files that are copied to the working directory, which is the default.

johnstairs replied:
Or maybe i'm missing something.

hansenms replied:
If we don't use the embedded files, we need to a) publish the SmartLauncher when we build and b) explicitly set to root folder when using the Startup in the E2E test. This approach means that we don't need to do any of those.

johnstairs replied:
I see, we use it from the E2E tests. Might not be so obvious that something under samples is used by the tests. 
Anyway, it's fine.

hansenms replied:
Yeah, the dual use is less than obvious. I could have kept it entirely in the tests, but I thought it might serve as an example as well.

hansenms commented on samples/apps/SmartLauncher/wwwroot/index.html:
```diff
@@ -0,0 +1,99 @@
+            launchContext.practioner = $('#practitioner').val();
+        }
+
+        var queryString = 'launch=' + encodeURIComponent(btoa(JSON.stringify(launchContext)));
```
We will do something on this at a later point per offline discussion.

johnstairs commented on samples/apps/SmartLauncher/wwwroot/index.html:
```diff
@@ -0,0 +1,99 @@
+    <div class="row">
+        <h2>Launch context</h2>
+        <div class="form-group col-xs-8">
+            <textarea class="form-control" rows="5" id="launchContext" >
```
Did you consider making this readonly, since edits to it are ignored?

johnstairs commented on samples/templates/default-azuredeploy.json:
```diff
@@ -40,6 +40,13 @@
         },
+        "allowedOrigins": {
+            "type": "array",
+            "defaultValue": [ "https://localhost:6001" ],
```
Reminder to make this empty.

hansenms commented on src/Microsoft.Health.Fhir.Api/Controllers/AadProxyController.cs:
```diff
@@ -0,0 +1,354 @@
+    /// Controller class enabling Azure Active Directory SMART on FHIR Proxy Capability
+    /// </summary>
+    [TypeFilter(typeof(AadProxyFeatureFilterAttribute))]
+    [Route("/AadProxy")]
```
We could....and should we store that path segment somewhere so that if and when we want to change it, we will do it in only one place. Do we have such a place?

johnstairs commented on src/Microsoft.Health.Fhir.Api/Controllers/AadProxyController.cs:
```diff
@@ -0,0 +1,354 @@
+        }
+
+        /// <summary>
+        /// Proxies a request to the Azure AD authorize endpoint.
```
Might be more accurate to say that it redirects to the AAD authorize endpoint with adjusted parameters.

johnstairs commented on test/Microsoft.Health.Fhir.Tests.E2E/SmartProxy/SmartProxyTestFixture.cs:
```diff
@@ -0,0 +1,74 @@
+
+        public SmartProxyTestFixture()
+        {
+            _environmentUrl = Environment.GetEnvironmentVariable("TestEnvironmentUrl");
```
This is not defined by default.

johnstairs commented on test/Microsoft.Health.Fhir.Tests.E2E/SmartProxy/SmartProxyTestFixture.cs:
```diff
@@ -0,0 +1,74 @@
+            WebServer = builder.Build();
+            WebServer.Start();
+
+            // _server = new TestServer(builder);
```
remvoe

johnstairs commented on src/Microsoft.Health.Fhir.Api/Controllers/AadProxyController.cs:
```diff
@@ -0,0 +1,354 @@
+                newState = launchStateParameters["s"].ToString();
+                compoundCode = Base64UrlEncoder.Encode(launchParameters.ToString(Newtonsoft.Json.Formatting.None));
+            }
+            catch
```
It would be better to return a 400 in this case

johnstairs commented on src/Microsoft.Health.Fhir.Api/Controllers/AadProxyController.cs:
```diff
@@ -0,0 +1,354 @@
+        private static bool IsAbsoluteUrl(string url)
+        {
+            Uri result;
+            return Uri.TryCreate(url, UriKind.Absolute, out result);
```
Syntactic sugar: you could write `return Uri.TryCreate(url, UriKind.Absolute, out _);`
